File: 238-product-of-array-except-self/product-of-array-except-self.cpp
class Solution {
public:
    vector<int> productExceptSelf(vector<int>& nums) {
        int n = nums.size();
        vector<int> res(n, 1);

        res[0] = 1;
        for(int i = 1; i < n; i++)
        {
            res[i] = res[i - 1] * nums[i - 1];
        }

        int rightSum = 1;
        for (int i = n - 1; i >= 0; i--) 
        {
            res[i] *= rightSum;
            rightSum *= nums[i];
        }

        return res;
    }
};

File: 238-product-of-array-except-self/product-of-array-except-self.cs
public class Solution {
    public int[] ProductExceptSelf(int[] nums) {
        int n = nums.Length;

        int[] res = new int[n];
        res[n - 1] = 1;

        for(int i = n - 2;i >= 0;i--)
        {
            res[i] = res[i + 1] * nums[i + 1];
        }

        int leftSum = 1;
        for(int k = 0;k < n;k++)
        {
            res[k] = res[k] * leftSum;
            leftSum *= nums[k];
        }

        return res;
    }
}